---

## **🚀 Projeto - Desempenho Acadêmico**

### **Inteligência Artificial e Previsões**:  
**Criar um modelo que consiga ler as informações dos estudantes e dizer automaticamente a classificação de suas notas: A, B, C, D, F**

Analisar o dataframe e com base na análise criar o modelo de Inteligência Artificial e Previsões.

## **📚 Origem dos Dados**
Este dataset foi disponibilizado por Rabie El Kharoua no Kaggle e está licenciado sob CC BY 4.0:  

> El Kharoua, R. (2024). Students Performance Dataset.  
> Disponível em: [Kaggle Dataset Link](https://www.kaggle.com/datasets/rabieelkharoua/students-performance-dataset).

---

## Passos do Projeto ##

Passo 1 - Importar o DataFrame

Passo 2 - Visualizar o DataFrame

Passo 3 - Corrigir erros no DataFrame, modelar e entender os dados

Passo 4 - Preparar o dataframe para a Inteligência Artificial

Passo 5 - Treinar a Inteligência Artificial

Passo 6 - Escolher o melhor modelo

Passo 7 - Usar o melhor modelo



---


Passo 1 - Importar o DataFrame

---

In [36]:
# Importando as bibliotecas necessárias
import pandas as pd

# Importando o dataset
df_dados_estudantes = pd.read_csv("dados_estudantes.csv", sep= ',')

# Verificando as primeiras linhas do dataset
display(df_dados_estudantes)


StudentID  Age  Gender  Ethnicity  ParentalEducation  StudyTimeWeekly  \
0          1001   17       1          0                  2        19.833723   
1          1002   18       0          0                  1        15.408756   
2          1003   15       0          2                  3         4.210570   
3          1004   17       1          0                  3        10.028829   
4          1005   17       1          0                  2         4.672495   
...         ...  ...     ...        ...                ...              ...   
2387       3388   18       1          0                  3        10.680555   
2388       3389   17       0          0                  1         7.583217   
2389       3390   16       1          0                  2         6.805500   
2390       3391   16       1          1                  0        12.416653   
2391       3392   16       1          0                  2        17.819907   

      Absences  Tutoring  ParentalSupport  Extracurricular  Sports  Music  \
0            7         1                2                0       0      1   
1            0         0                1                0       0      0   
2           26         0                2                0       0      0   
3           14         0                3                1       0      0   
4           17         1                3                0       0      0   
...        ...       ...              ...              ...     ...    ...   
2387         2         0                4                1       0      0   
2388         4         1                4                0       1      0   
2389        20         0                2                0       0      0   
2390        17         0                2                0       1      1   
2391        13         0                2                0       0      0   

      Volunteering       GPA  GradeClass  
0                0  2.929196         2.0  
1                0  3.042915         1.0  
2                0  0.112602         4.0  
3                0  2.054218         3.0  
4                0  1.288061         4.0  
...            ...       ...         ...  
2387             0  3.455509         0.0  
2388             0  3.279150         4.0  
2389             1  1.142333         2.0  
2390             0  1.803297         1.0  
2391             1  2.140014         1.0  

[2392 rows x 15 columns]

In [37]:
traducao_colunas = {
    "StudentID": "ID_Aluno",
    "Age": "Idade",
    "Gender": "Genero",
    "Ethnicity": "Etnia",
    "ParentalEducation": "Escolaridade_Pais",
    "StudyTimeWeekly": "Horas_Estudo_Semanais",
    "Absences": "Faltas",
    "Tutoring": "Aulas_Particulares",
    "ParentalSupport": "Apoio_Pais",
    "Extracurricular": "Atividades_Extracurriculares",
    "Sports": "Esportes",
    "Music": "Musica",
    "Volunteering": "Voluntariado",
    "GPA": "Media_Notas",
    "GradeClass": "Classificacao_Notas"
}

# Renomear as colunas
df_dados_estudantes = df_dados_estudantes.rename(columns=traducao_colunas)

df_dados_estudantes.head()


ID_Aluno  Idade  Genero  Etnia  Escolaridade_Pais  Horas_Estudo_Semanais  \
0      1001     17       1      0                  2              19.833723   
1      1002     18       0      0                  1              15.408756   
2      1003     15       0      2                  3               4.210570   
3      1004     17       1      0                  3              10.028829   
4      1005     17       1      0                  2               4.672495   

   Faltas  Aulas_Particulares  Apoio_Pais  Atividades_Extracurriculares  \
0       7                   1           2                             0   
1       0                   0           1                             0   
2      26                   0           2                             0   
3      14                   0           3                             1   
4      17                   1           3                             0   

   Esportes  Musica  Voluntariado  Media_Notas  Classificacao_Notas  
0         0       1             0     2.929196                  2.0  
1         0       0             0     3.042915                  1.0  
2         0       0             0     0.112602                  4.0  
3         0       0             0     2.054218                  3.0  
4         0       0             0     1.288061                  4.0

---

Passo 2 - Visualizar o DataFrame

---

In [38]:
# Usando shape para conferir as dimensões do dataframe
df_dados_estudantes.shape

(2392, 15)

In [39]:
# Conferindo os tipos de dados utilizados no dataframe com dtypes
df_dados_estudantes.dtypes

ID_Aluno                          int64
Idade                             int64
Genero                            int64
Etnia                             int64
Escolaridade_Pais                 int64
Horas_Estudo_Semanais           float64
Faltas                            int64
Aulas_Particulares                int64
Apoio_Pais                        int64
Atividades_Extracurriculares      int64
Esportes                          int64
Musica                            int64
Voluntariado                      int64
Media_Notas                     float64
Classificacao_Notas             float64
dtype: object

In [40]:
# Visualizando as informaçoes do dataframe com info
df_dados_estudantes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2392 entries, 0 to 2391
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   ID_Aluno                      2392 non-null   int64  
 1   Idade                         2392 non-null   int64  
 2   Genero                        2392 non-null   int64  
 3   Etnia                         2392 non-null   int64  
 4   Escolaridade_Pais             2392 non-null   int64  
 5   Horas_Estudo_Semanais         2392 non-null   float64
 6   Faltas                        2392 non-null   int64  
 7   Aulas_Particulares            2392 non-null   int64  
 8   Apoio_Pais                    2392 non-null   int64  
 9   Atividades_Extracurriculares  2392 non-null   int64  
 10  Esportes                      2392 non-null   int64  
 11  Musica                        2392 non-null   int64  
 12  Voluntariado                  2392 non-null   int64  
 13  Med

In [41]:
# Conferir se existe valores ausentes no dataframe
df_dados_estudantes.isna().sum()

ID_Aluno                        0
Idade                           0
Genero                          0
Etnia                           0
Escolaridade_Pais               0
Horas_Estudo_Semanais           0
Faltas                          0
Aulas_Particulares              0
Apoio_Pais                      0
Atividades_Extracurriculares    0
Esportes                        0
Musica                          0
Voluntariado                    0
Media_Notas                     0
Classificacao_Notas             0
dtype: int64

In [42]:
# Por fim fazer uma análise descritiva do dataframe
df_dados_estudantes.describe()

ID_Aluno        Idade       Genero        Etnia  Escolaridade_Pais  \
count  2392.000000  2392.000000  2392.000000  2392.000000        2392.000000   
mean   2196.500000    16.468645     0.510870     0.877508           1.746237   
std     690.655244     1.123798     0.499986     1.028476           1.000411   
min    1001.000000    15.000000     0.000000     0.000000           0.000000   
25%    1598.750000    15.000000     0.000000     0.000000           1.000000   
50%    2196.500000    16.000000     1.000000     0.000000           2.000000   
75%    2794.250000    17.000000     1.000000     2.000000           2.000000   
max    3392.000000    18.000000     1.000000     3.000000           4.000000   

       Horas_Estudo_Semanais       Faltas  Aulas_Particulares   Apoio_Pais  \
count            2392.000000  2392.000000         2392.000000  2392.000000   
mean                9.771992    14.541388            0.301421     2.122074   
std                 5.652774     8.467417            0.458971     1.122813   
min                 0.001057     0.000000            0.000000     0.000000   
25%                 5.043079     7.000000            0.000000     1.000000   
50%                 9.705363    15.000000            0.000000     2.000000   
75%                14.408410    22.000000            1.000000     3.000000   
max                19.978094    29.000000            1.000000     4.000000   

       Atividades_Extracurriculares     Esportes       Musica  Voluntariado  \
count                   2392.000000  2392.000000  2392.000000   2392.000000   
mean                       0.383361     0.303512     0.196906      0.157191   
std                        0.486307     0.459870     0.397744      0.364057   
min                        0.000000     0.000000     0.000000      0.000000   
25%                        0.000000     0.000000     0.000000      0.000000   
50%                        0.000000     0.000000     0.000000      0.000000   
75%                        1.000000     1.000000     0.000000      0.000000   
max                        1.000000     1.000000     1.000000      1.000000   

       Media_Notas  Classificacao_Notas  
count  2392.000000          2392.000000  
mean      1.906186             2.983696  
std       0.915156             1.233908  
min       0.000000             0.000000  
25%       1.174803             2.000000  
50%       1.893393             4.000000  
75%       2.622216             4.000000  
max       4.000000             4.000000

---

Passo 3 - Corrigir erros no DataFrame, modelar e entender os dados

---

Para seguir os padrões apresentados na Descrição das Colunas do DataFrame, irei alterar o tipo de dados de algumas colunas, seguindo o padrão descrito abaixo:

## **📊 Descrição das Colunas do DataFrame**

### **Identificação do Estudante**  
- **ID_Aluno**: Identificador único para cada aluno (de 1001 a 3392).

### **👥 Detalhes Demográficos**  
- **Idade**: Idade dos estudantes, variando de 15 a 18 anos.  
- **Genero**: Gênero do estudante:
  - 0: Masculino
  - 1: Feminino  
- **Etnia**: Etnia dos estudantes:
  - 0: Caucasiano
  - 1: Afrodescendente
  - 2: Asiático
  - 3: Outra

### **👨‍👩‍👧 Educação dos Pais**  
- **Escolaridade_Pais**: Nível de escolaridade dos pais:
  - 0: Nenhum
  - 1: Ensino Médio
  - 2: Alguma Faculdade
  - 3: Graduação
  - 4: Pós-Graduação ou Superior

### **📚 Hábitos de Estudo**  
- **Horas_Estudo_Semanais**: Tempo de estudo semanal em horas (0 a 20).  
- **Faltas**: Número de faltas no ano letivo (0 a 30).  
- **Aulas_Particulares**: Participação em aulas particulares:
  - 0: Não
  - 1: Sim

### **👨‍👩‍👧 Envolvimento Parental**  
- **Apoio_Pais**: Nível de apoio dos pais:
  - 0: Nenhum
  - 1: Baixo
  - 2: Moderado
  - 3: Alto
  - 4: Muito Alto

### **🏫 Atividades Extracurriculares**  
- **Atividades_Extracurriculares**: Participação em atividades extracurriculares:
  - 0: Não
  - 1: Sim  
- **Esportes**: Participação em esportes:
  - 0: Não
  - 1: Sim  
- **Musica**: Participação em atividades musicais:
  - 0: Não
  - 1: Sim  
- **Voluntariado**: Participação em atividades de voluntariado:
  - 0: Não
  - 1: Sim

### **📈 Desempenho Acadêmico**  
- **Media_Notas**: Média das notas (GPA), variando de 2.0 a 4.0.  
- **Classificacao_Notas**: Classificação do aluno com base na média das notas:
  - 0: 'A' (Média >= 3.5)
  - 1: 'B' (3.0 <= Média < 3.5)
  - 2: 'C' (2.5 <= Média < 3.0)
  - 3: 'D' (2.0 <= Média < 2.5)
  - 4: 'F' (Média < 2.0)

---

In [43]:
# Alterando a coluna Media_Notas para float com 1 casa decimal
df_dados_estudantes[["Media_Notas"]] = df_dados_estudantes[["Media_Notas"]].astype(float).round(1)


In [44]:
# Alterando as colunas Horas_Estudo_Semanais e Classificacao_Notas para Int
df_dados_estudantes[["Horas_Estudo_Semanais", 
                     "Classificacao_Notas"]] = df_dados_estudantes[["Horas_Estudo_Semanais", 
                                                                    "Classificacao_Notas"]].astype(int)

# Visualizando a alteração das 3 colunas
df_dados_estudantes.dtypes

ID_Aluno                          int64
Idade                             int64
Genero                            int64
Etnia                             int64
Escolaridade_Pais                 int64
Horas_Estudo_Semanais             int64
Faltas                            int64
Aulas_Particulares                int64
Apoio_Pais                        int64
Atividades_Extracurriculares      int64
Esportes                          int64
Musica                            int64
Voluntariado                      int64
Media_Notas                     float64
Classificacao_Notas               int64
dtype: object

In [45]:
# Visualizando o dataframe após as alterações
df_dados_estudantes.head()

ID_Aluno  Idade  Genero  Etnia  Escolaridade_Pais  Horas_Estudo_Semanais  \
0      1001     17       1      0                  2                     19   
1      1002     18       0      0                  1                     15   
2      1003     15       0      2                  3                      4   
3      1004     17       1      0                  3                     10   
4      1005     17       1      0                  2                      4   

   Faltas  Aulas_Particulares  Apoio_Pais  Atividades_Extracurriculares  \
0       7                   1           2                             0   
1       0                   0           1                             0   
2      26                   0           2                             0   
3      14                   0           3                             1   
4      17                   1           3                             0   

   Esportes  Musica  Voluntariado  Media_Notas  Classificacao_Notas  
0         0       1             0          2.9                    2  
1         0       0             0          3.0                    1  
2         0       0             0          0.1                    4  
3         0       0             0          2.1                    3  
4         0       0             0          1.3                    4

In [46]:
# O ID_Aluno está sendo contando de 1001 à 3392, para alterar a contagem para 1 à 2392 usarei o método apply lambda para 
# fazer essa alteração.

# Alterando os valores do ID_Aluno para contar do número 1 em diante
df_dados_estudantes["ID_Aluno"] = df_dados_estudantes["ID_Aluno"].apply(lambda x: x - 1000)
     

In [47]:
# Para alterar os valores numéricos pelos seus respectivos valores informados na descrição, usarei o método replace 
# para reverter os números cadastrados em seus devidos valores informados.

# Alterando os valores da descrição da coluna Gênero
df_dados_estudantes["Genero"] = df_dados_estudantes["Genero"].replace({
     0: "Masculino", 
     1: "Feminino"
})


In [48]:
# Alterando os valores da descrição da coluna Etnia
df_dados_estudantes["Etnia"] = df_dados_estudantes["Etnia"].replace({
    0: "Caucasiano", 
    1: "Afrodescendente", 
    2: "Asiático", 
    3: "Outra"
})


In [49]:
# Alterando os valores da descrição da coluna Escolaridade dos Pais
df_dados_estudantes["Escolaridade_Pais"] = df_dados_estudantes["Escolaridade_Pais"].replace({
     0: "Nenhum", 
     1: "Ensino Médio", 
     2: "Alguma Faculdade", 
     3: "Graduação", 
     4: "Pós-Graduação ou Superior"
})


In [50]:
# Alterando os valores da descrição da coluna Aulas Particulares
df_dados_estudantes["Aulas_Particulares"] = df_dados_estudantes["Aulas_Particulares"].replace({
    0: "Não", 
    1: "Sim"
})


In [51]:
# Alterando os valores da descrição da coluna Apoio dos Pais
df_dados_estudantes["Apoio_Pais"] = df_dados_estudantes["Apoio_Pais"].replace({
    0: "Nenhum", 
    1: "Baixo", 
    2: "Moderado", 
    3: "Alto", 
    4: "Muito Alto"
})


In [52]:
# Alterando os valores da descrição da coluna Atividades Extracurriculares
df_dados_estudantes["Atividades_Extracurriculares"] = df_dados_estudantes["Atividades_Extracurriculares"].replace({
    0: "Não",
    1: "Sim"
})

In [53]:
# Alterando os valores da descrição da coluna Esportes
df_dados_estudantes["Esportes"] = df_dados_estudantes["Esportes"].replace({
     0: "Não", 
     1: "Sim"
})


In [54]:
# Alterando os valores da descrição da coluna Musica
df_dados_estudantes["Musica"] = df_dados_estudantes["Musica"].replace({
    0: "Não", 
    1: "Sim"
})


In [55]:
# Alterando os valores da descrição da coluna Voluntariado
df_dados_estudantes["Voluntariado"] = df_dados_estudantes["Voluntariado"].replace({
    0: "Não", 
    1: "Sim"
})


In [56]:
# Alterando os valores da descrição da coluna Classificacao_Notas
df_dados_estudantes["Classificacao_Notas"] = df_dados_estudantes["Classificacao_Notas"].replace({
    0: "A", 
    1: "B", 
    2: "C", 
    3: "D", 
    4: "F"
})

In [57]:
# Com a alteração de todos os valores necessários para suas devidas descrições, é hora de visualizar o resultado em 
# um novo dataframe.

#Visualização do novo dataframe após a alteração das descrições das colunas
novo_df_estudantes = df_dados_estudantes

display(novo_df_estudantes)


ID_Aluno  Idade     Genero            Etnia Escolaridade_Pais  \
0            1     17   Feminino       Caucasiano  Alguma Faculdade   
1            2     18  Masculino       Caucasiano      Ensino Médio   
2            3     15  Masculino         Asiático         Graduação   
3            4     17   Feminino       Caucasiano         Graduação   
4            5     17   Feminino       Caucasiano  Alguma Faculdade   
...        ...    ...        ...              ...               ...   
2387      2388     18   Feminino       Caucasiano         Graduação   
2388      2389     17  Masculino       Caucasiano      Ensino Médio   
2389      2390     16   Feminino       Caucasiano  Alguma Faculdade   
2390      2391     16   Feminino  Afrodescendente            Nenhum   
2391      2392     16   Feminino       Caucasiano  Alguma Faculdade   

      Horas_Estudo_Semanais  Faltas Aulas_Particulares  Apoio_Pais  \
0                        19       7                Sim    Moderado   
1                        15       0                Não       Baixo   
2                         4      26                Não    Moderado   
3                        10      14                Não        Alto   
4                         4      17                Sim        Alto   
...                     ...     ...                ...         ...   
2387                     10       2                Não  Muito Alto   
2388                      7       4                Sim  Muito Alto   
2389                      6      20                Não    Moderado   
2390                     12      17                Não    Moderado   
2391                     17      13                Não    Moderado   

     Atividades_Extracurriculares Esportes Musica Voluntariado  Media_Notas  \
0                             Não      Não    Sim          Não          2.9   
1                             Não      Não    Não          Não          3.0   
2                             Não      Não    Não          Não          0.1   
3                             Sim      Não    Não          Não          2.1   
4                             Não      Não    Não          Não          1.3   
...                           ...      ...    ...          ...          ...   
2387                          Sim      Não    Não          Não          3.5   
2388                          Não      Sim    Não          Não          3.3   
2389                          Não      Não    Não          Sim          1.1   
2390                          Não      Sim    Sim          Não          1.8   
2391                          Não      Não    Não          Sim          2.1   

     Classificacao_Notas  
0                      C  
1                      B  
2                      F  
3                      D  
4                      F  
...                  ...  
2387                   A  
2388                   F  
2389                   C  
2390                   B  
2391                   B  

[2392 rows x 15 columns]

Ao analisar o novo DataFrame, observei que existem inconsistências nos valores em relação a coluna Media_Notas e Classificacao_Notas. Para corrigir esse problema e não haver divergências entre as análises, irei corrigir os valores usando o método apply com lambda de acordo com as informações descritas na Descrição do Desempenho Acadêmico.

In [58]:
# Aplicando o lambda para corrigir os valores da Classificacao_Notas em relação a Média_Notas
novo_df_estudantes["Classificacao_Notas_Lambda"] = novo_df_estudantes["Media_Notas"].apply(lambda x: 
                                                                                           "A" if x >= 3.5 else 
                                                                                           "B" if x >= 3.0 else 
                                                                                           "C" if x >= 2.5 else 
                                                                                           "D" if x >= 2.0 else 
                                                                                           "F")

display(novo_df_estudantes)

ID_Aluno  Idade     Genero            Etnia Escolaridade_Pais  \
0            1     17   Feminino       Caucasiano  Alguma Faculdade   
1            2     18  Masculino       Caucasiano      Ensino Médio   
2            3     15  Masculino         Asiático         Graduação   
3            4     17   Feminino       Caucasiano         Graduação   
4            5     17   Feminino       Caucasiano  Alguma Faculdade   
...        ...    ...        ...              ...               ...   
2387      2388     18   Feminino       Caucasiano         Graduação   
2388      2389     17  Masculino       Caucasiano      Ensino Médio   
2389      2390     16   Feminino       Caucasiano  Alguma Faculdade   
2390      2391     16   Feminino  Afrodescendente            Nenhum   
2391      2392     16   Feminino       Caucasiano  Alguma Faculdade   

      Horas_Estudo_Semanais  Faltas Aulas_Particulares  Apoio_Pais  \
0                        19       7                Sim    Moderado   
1                        15       0                Não       Baixo   
2                         4      26                Não    Moderado   
3                        10      14                Não        Alto   
4                         4      17                Sim        Alto   
...                     ...     ...                ...         ...   
2387                     10       2                Não  Muito Alto   
2388                      7       4                Sim  Muito Alto   
2389                      6      20                Não    Moderado   
2390                     12      17                Não    Moderado   
2391                     17      13                Não    Moderado   

     Atividades_Extracurriculares Esportes Musica Voluntariado  Media_Notas  \
0                             Não      Não    Sim          Não          2.9   
1                             Não      Não    Não          Não          3.0   
2                             Não      Não    Não          Não          0.1   
3                             Sim      Não    Não          Não          2.1   
4                             Não      Não    Não          Não          1.3   
...                           ...      ...    ...          ...          ...   
2387                          Sim      Não    Não          Não          3.5   
2388                          Não      Sim    Não          Não          3.3   
2389                          Não      Não    Não          Sim          1.1   
2390                          Não      Sim    Sim          Não          1.8   
2391                          Não      Não    Não          Sim          2.1   

     Classificacao_Notas Classificacao_Notas_Lambda  
0                      C                          C  
1                      B                          B  
2                      F                          F  
3                      D                          D  
4                      F                          F  
...                  ...                        ...  
2387                   A                          A  
2388                   F                          B  
2389                   C                          F  
2390                   B                          F  
2391                   B                          D  

[2392 rows x 16 columns]

Como eu havia notado, os valores estavam divergentes. Isso pode ser observado no dataframe comparando-se os valores entre Classificacao_Notas e Classificacao_Notas_Lambda que eu criei.

In [59]:
# Exluindo a coluna Classificacao_Notas (com os valores errados)
novo_df_estudantes = novo_df_estudantes.drop(columns=["Classificacao_Notas"])

novo_df_estudantes.head()

ID_Aluno  Idade     Genero       Etnia Escolaridade_Pais  \
0         1     17   Feminino  Caucasiano  Alguma Faculdade   
1         2     18  Masculino  Caucasiano      Ensino Médio   
2         3     15  Masculino    Asiático         Graduação   
3         4     17   Feminino  Caucasiano         Graduação   
4         5     17   Feminino  Caucasiano  Alguma Faculdade   

   Horas_Estudo_Semanais  Faltas Aulas_Particulares Apoio_Pais  \
0                     19       7                Sim   Moderado   
1                     15       0                Não      Baixo   
2                      4      26                Não   Moderado   
3                     10      14                Não       Alto   
4                      4      17                Sim       Alto   

  Atividades_Extracurriculares Esportes Musica Voluntariado  Media_Notas  \
0                          Não      Não    Sim          Não          2.9   
1                          Não      Não    Não          Não          3.0   
2                          Não      Não    Não          Não          0.1   
3                          Sim      Não    Não          Não          2.1   
4                          Não      Não    Não          Não          1.3   

  Classificacao_Notas_Lambda  
0                          C  
1                          B  
2                          F  
3                          D  
4                          F

In [60]:
# Renomeando a coluna Classificacao_Notas_Lambda para o nome original Classificacao_Notas
novo_df_estudantes = novo_df_estudantes.rename(columns={"Classificacao_Notas_Lambda": "Classificacao_Notas"})

novo_df_estudantes.head()

ID_Aluno  Idade     Genero       Etnia Escolaridade_Pais  \
0         1     17   Feminino  Caucasiano  Alguma Faculdade   
1         2     18  Masculino  Caucasiano      Ensino Médio   
2         3     15  Masculino    Asiático         Graduação   
3         4     17   Feminino  Caucasiano         Graduação   
4         5     17   Feminino  Caucasiano  Alguma Faculdade   

   Horas_Estudo_Semanais  Faltas Aulas_Particulares Apoio_Pais  \
0                     19       7                Sim   Moderado   
1                     15       0                Não      Baixo   
2                      4      26                Não   Moderado   
3                     10      14                Não       Alto   
4                      4      17                Sim       Alto   

  Atividades_Extracurriculares Esportes Musica Voluntariado  Media_Notas  \
0                          Não      Não    Sim          Não          2.9   
1                          Não      Não    Não          Não          3.0   
2                          Não      Não    Não          Não          0.1   
3                          Sim      Não    Não          Não          2.1   
4                          Não      Não    Não          Não          1.3   

  Classificacao_Notas  
0                   C  
1                   B  
2                   F  
3                   D  
4                   F

Agora que todas as colunas e seus valores foram remodelados e corrigidos, temos uma visão melhorada e precisa das informações, irei preparar o dataframe para a inteligência artifical.

In [61]:
novo_df_estudantes

ID_Aluno  Idade     Genero            Etnia Escolaridade_Pais  \
0            1     17   Feminino       Caucasiano  Alguma Faculdade   
1            2     18  Masculino       Caucasiano      Ensino Médio   
2            3     15  Masculino         Asiático         Graduação   
3            4     17   Feminino       Caucasiano         Graduação   
4            5     17   Feminino       Caucasiano  Alguma Faculdade   
...        ...    ...        ...              ...               ...   
2387      2388     18   Feminino       Caucasiano         Graduação   
2388      2389     17  Masculino       Caucasiano      Ensino Médio   
2389      2390     16   Feminino       Caucasiano  Alguma Faculdade   
2390      2391     16   Feminino  Afrodescendente            Nenhum   
2391      2392     16   Feminino       Caucasiano  Alguma Faculdade   

      Horas_Estudo_Semanais  Faltas Aulas_Particulares  Apoio_Pais  \
0                        19       7                Sim    Moderado   
1                        15       0                Não       Baixo   
2                         4      26                Não    Moderado   
3                        10      14                Não        Alto   
4                         4      17                Sim        Alto   
...                     ...     ...                ...         ...   
2387                     10       2                Não  Muito Alto   
2388                      7       4                Sim  Muito Alto   
2389                      6      20                Não    Moderado   
2390                     12      17                Não    Moderado   
2391                     17      13                Não    Moderado   

     Atividades_Extracurriculares Esportes Musica Voluntariado  Media_Notas  \
0                             Não      Não    Sim          Não          2.9   
1                             Não      Não    Não          Não          3.0   
2                             Não      Não    Não          Não          0.1   
3                             Sim      Não    Não          Não          2.1   
4                             Não      Não    Não          Não          1.3   
...                           ...      ...    ...          ...          ...   
2387                          Sim      Não    Não          Não          3.5   
2388                          Não      Sim    Não          Não          3.3   
2389                          Não      Não    Não          Sim          1.1   
2390                          Não      Sim    Sim          Não          1.8   
2391                          Não      Não    Não          Sim          2.1   

     Classificacao_Notas  
0                      C  
1                      B  
2                      F  
3                      D  
4                      F  
...                  ...  
2387                   A  
2388                   B  
2389                   F  
2390                   F  
2391                   D  

[2392 rows x 15 columns]

---

Passo 4 - Preparar o dataframe para a Inteligência Artifical

---

In [62]:
# Visualizar as informações do dataframe
novo_df_estudantes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2392 entries, 0 to 2391
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   ID_Aluno                      2392 non-null   int64  
 1   Idade                         2392 non-null   int64  
 2   Genero                        2392 non-null   object 
 3   Etnia                         2392 non-null   object 
 4   Escolaridade_Pais             2392 non-null   object 
 5   Horas_Estudo_Semanais         2392 non-null   int64  
 6   Faltas                        2392 non-null   int64  
 7   Aulas_Particulares            2392 non-null   object 
 8   Apoio_Pais                    2392 non-null   object 
 9   Atividades_Extracurriculares  2392 non-null   object 
 10  Esportes                      2392 non-null   object 
 11  Musica                        2392 non-null   object 
 12  Voluntariado                  2392 non-null   object 
 13  Med

In [ ]:
# Codificar as colunas para tipo numérico para a Inteligência Artificial

from sklearn.preprocessing import LabelEncoder

# Codificar coluna Genero
codificador_genero = LabelEncoder()
novo_df_estudantes["Genero"] = codificador_genero.fit_transform(novo_df_estudantes["Genero"])

# Codificar coluna Etnia
codificado_etnia = LabelEncoder()
novo_df_estudantes["Etnia"] = codificado_etnia.fit_transform(novo_df_estudantes["Etnia"])

# Codificar coluna Escolaridade_Pais
codificador_escola_pais = LabelEncoder()
novo_df_estudantes["Escolaridade_Pais"] = codificador_escola_pais.fit_transform(novo_df_estudantes["Escolaridade_Pais"])

# Codificar coluna Aulas Particulares
codificador_particulares = LabelEncoder()
novo_df_estudantes["Aulas_Particulares"] = codificador_particulares.fit_transform(novo_df_estudantes["Aulas_Particulares"])

# Codificar coluna Apoio_Pais
codificador_apoio = LabelEncoder()
novo_df_estudantes["Apoio_Pais"] = codificador_apoio.fit_transform(novo_df_estudantes["Apoio_Pais"])

# Codificar coluna Ativiadades Extracurriculares
codificador_atividade = LabelEncoder()
novo_df_estudantes["Atividades_Extracurriculares"] = codificador_atividade.fit_transform(novo_df_estudantes["Atividades_Extracurriculares"])

# Codificar coluna Esportes
codificador_esporte = LabelEncoder()
novo_df_estudantes["Esportes"] = codificador_esporte.fit_transform(novo_df_estudantes["Esportes"])

# Codificar coluna Musica
codificador_musica = LabelEncoder()
novo_df_estudantes["Musica"] = codificador_musica.fit_transform(novo_df_estudantes["Musica"])

# Codificar coluna Voluntariado
codificador_voluntariado = LabelEncoder()
novo_df_estudantes["Voluntariado"] = codificador_voluntariado.fit_transform(novo_df_estudantes["Voluntariado"])

novo_df_estudantes.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2392 entries, 0 to 2391
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   ID_Aluno                      2392 non-null   int64  
 1   Idade                         2392 non-null   int64  
 2   Genero                        2392 non-null   int64  
 3   Etnia                         2392 non-null   int64  
 4   Escolaridade_Pais             2392 non-null   int64  
 5   Horas_Estudo_Semanais         2392 non-null   int64  
 6   Faltas                        2392 non-null   int64  
 7   Aulas_Particulares            2392 non-null   int64  
 8   Apoio_Pais                    2392 non-null   int64  
 9   Atividades_Extracurriculares  2392 non-null   int64  
 10  Esportes                      2392 non-null   int64  
 11  Musica                        2392 non-null   int64  
 12  Voluntariado                  2392 non-null   int64  
 13  Med

In [64]:
# Visualização do dataframe após a codificação
display(novo_df_estudantes)


ID_Aluno  Idade  Genero  Etnia  Escolaridade_Pais  \
0            1     17       0      2                  0   
1            2     18       1      2                  1   
2            3     15       1      1                  2   
3            4     17       0      2                  2   
4            5     17       0      2                  0   
...        ...    ...     ...    ...                ...   
2387      2388     18       0      2                  2   
2388      2389     17       1      2                  1   
2389      2390     16       0      2                  0   
2390      2391     16       0      0                  3   
2391      2392     16       0      2                  0   

      Horas_Estudo_Semanais  Faltas  Aulas_Particulares  Apoio_Pais  \
0                        19       7                   1           2   
1                        15       0                   0           1   
2                         4      26                   0           2   
3                        10      14                   0           0   
4                         4      17                   1           0   
...                     ...     ...                 ...         ...   
2387                     10       2                   0           3   
2388                      7       4                   1           3   
2389                      6      20                   0           2   
2390                     12      17                   0           2   
2391                     17      13                   0           2   

      Atividades_Extracurriculares  Esportes  Musica  Voluntariado  \
0                                0         0       1             0   
1                                0         0       0             0   
2                                0         0       0             0   
3                                1         0       0             0   
4                                0         0       0             0   
...                            ...       ...     ...           ...   
2387                             1         0       0             0   
2388                             0         1       0             0   
2389                             0         0       0             1   
2390                             0         1       1             0   
2391                             0         0       0             1   

      Media_Notas Classificacao_Notas  
0             2.9                   C  
1             3.0                   B  
2             0.1                   F  
3             2.1                   D  
4             1.3                   F  
...           ...                 ...  
2387          3.5                   A  
2388          3.3                   B  
2389          1.1                   F  
2390          1.8                   F  
2391          2.1                   D  

[2392 rows x 15 columns]

In [65]:
# Preparar o dataframe para o treino

# y é a coluna do dataframe a ser prevista
y = novo_df_estudantes["Classificacao_Notas"]

# x são as colunas do dataframe que serão usadas para a previsão
x = novo_df_estudantes.drop(columns=["Classificacao_Notas", "ID_Aluno"])

# Separar os dados de treino e os dados de teste
from sklearn.model_selection import train_test_split

x_treino, x_teste, y_treino, y_teste = train_test_split(x, y, test_size=0.7)


---

Passo 5 - Treinar a Inteligência Artificial

---

In [66]:
# Criar o medelo -> Classificação da Nota: A, B, C, D, F

# Arvore de Decisão -> RandonForest
# Nearest Neighbors -> KNN -> Vizinhos Próximos

# Importar a Inteligência Artificial (IA)
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier

# Cria a IA
modelo_arvore_decisao = RandomForestClassifier()
modelo_knn = KNeighborsClassifier()

# Treinar a IA
modelo_arvore_decisao.fit(x_treino, y_treino)
modelo_knn.fit(x_treino, y_treino)

KNeighborsClassifier()

---

Passo 6 - Escolher o melhor modelo

---

In [67]:
# Testar os modelos
previsao_arvore_decisao = modelo_arvore_decisao.predict(x_teste)
previsao_knn = modelo_knn.predict(x_teste)

# Verificar Acurácia
from sklearn.metrics import accuracy_score

display(accuracy_score(y_teste, previsao_arvore_decisao) *100)
display(accuracy_score(y_teste, previsao_knn) * 100)

98.32835820895522

71.34328358208955

---

Passo 7 - Usar melhor modelo

---

In [68]:
# Melhor modelo é Modelo Arvore de Decisão

#Importar os novos estudantes para fazer a previsão
df_novos_dados = pd.read_csv("novos_dados_estudantes.csv")

display(df_novos_dados)

nova_previsao = modelo_arvore_decisao.predict(df_novos_dados)

display(nova_previsao)

Idade  Genero  Etnia  Escolaridade_Pais  Horas_Estudo_Semanais  Faltas  \
0       18       0      0                  1               5.344198      26   
1       15       0      2                  1              10.182268      21   
2       15       1      1                  1               2.949078       3   
3       17       1      2                  1              12.233114      21   
4       17       1      3                  2              11.314946      23   
..     ...     ...    ...                ...                    ...     ...   
253     18       1      0                  0               7.945826      15   
254     17       1      0                  2               6.736954      27   
255     15       1      0                  2              11.389648       3   
256     17       1      2                  2              17.222270      11   
257     16       0      1                  1               1.566764      10   

     Aulas_Particulares  Apoio_Pais  Atividades_Extracurriculares  Esportes  \
0                     1           1                             0         0   
1                     1           1                             0         1   
2                     1           1                             1         1   
3                     1           3                             1         1   
4                     1           1                             0         0   
..                  ...         ...                           ...       ...   
253                   1           0                             0         1   
254                   0           4                             0         0   
255                   1           3                             1         0   
256                   0           3                             0         0   
257                   0           3                             0         0   

     Musica  Voluntariado  Media_Notas  
0         0             0     0.218285  
1         0             0     1.067764  
2         0             0     3.018906  
3         0             0     1.526868  
4         0             0     1.058287  
..      ...           ...          ...  
253       0             0     1.253452  
254       0             0     0.534509  
255       0             0     3.191046  
256       1             0     2.815421  
257       0             0     2.157122  

[258 rows x 13 columns]

array(['F', 'F', 'B', 'F', 'F', 'C', 'C', 'C', 'A', 'C', 'C', 'F', 'D',
       'C', 'C', 'D', 'F', 'F', 'F', 'F', 'F', 'F', 'C', 'F', 'F', 'D',
       'D', 'D', 'B', 'B', 'F', 'F', 'F', 'B', 'C', 'F', 'C', 'F', 'F',
       'C', 'F', 'F', 'F', 'D', 'F', 'F', 'F', 'C', 'B', 'A', 'B', 'F',
       'D', 'F', 'F', 'F', 'F', 'A', 'D', 'B', 'F', 'B', 'A', 'C', 'A',
       'C', 'D', 'F', 'D', 'B', 'A', 'D', 'B', 'D', 'D', 'F', 'C', 'D',
       'B', 'F', 'F', 'D', 'D', 'C', 'A', 'D', 'F', 'C', 'C', 'F', 'F',
       'F', 'B', 'D', 'F', 'C', 'F', 'F', 'B', 'F', 'A', 'C', 'D', 'F',
       'C', 'D', 'D', 'F', 'F', 'F', 'D', 'C', 'F', 'F', 'C', 'F', 'C',
       'B', 'B', 'F', 'D', 'B', 'B', 'F', 'B', 'C', 'F', 'F', 'F', 'D',
       'D', 'D', 'C', 'F', 'B', 'F', 'F', 'F', 'F', 'F', 'B', 'D', 'F',
       'D', 'F', 'F', 'F', 'F', 'C', 'F', 'B', 'F', 'D', 'F', 'F', 'C',
       'F', 'F', 'F', 'D', 'D', 'C', 'F', 'F', 'F', 'F', 'F', 'B', 'F',
       'F', 'F', 'D', 'F', 'C', 'B', 'F', 'F', 'F', 'B', 'F', 'C

In [69]:
# Análise exploratória inicial da classificação das notas

# Contagem das notas da classificação

df_novos_dados["Classificacao_Notas"] = nova_previsao

display(df_novos_dados["Classificacao_Notas"].value_counts())

# Proporção das classificações
display(df_novos_dados["Classificacao_Notas"].value_counts(normalize=True).map("{:.1%}".format))

display(df_novos_dados)

Classificacao_Notas
F    126
D     48
C     39
B     36
A      9
Name: count, dtype: int64

Classificacao_Notas
F    48.8%
D    18.6%
C    15.1%
B    14.0%
A     3.5%
Name: proportion, dtype: object

Idade  Genero  Etnia  Escolaridade_Pais  Horas_Estudo_Semanais  Faltas  \
0       18       0      0                  1               5.344198      26   
1       15       0      2                  1              10.182268      21   
2       15       1      1                  1               2.949078       3   
3       17       1      2                  1              12.233114      21   
4       17       1      3                  2              11.314946      23   
..     ...     ...    ...                ...                    ...     ...   
253     18       1      0                  0               7.945826      15   
254     17       1      0                  2               6.736954      27   
255     15       1      0                  2              11.389648       3   
256     17       1      2                  2              17.222270      11   
257     16       0      1                  1               1.566764      10   

     Aulas_Particulares  Apoio_Pais  Atividades_Extracurriculares  Esportes  \
0                     1           1                             0         0   
1                     1           1                             0         1   
2                     1           1                             1         1   
3                     1           3                             1         1   
4                     1           1                             0         0   
..                  ...         ...                           ...       ...   
253                   1           0                             0         1   
254                   0           4                             0         0   
255                   1           3                             1         0   
256                   0           3                             0         0   
257                   0           3                             0         0   

     Musica  Voluntariado  Media_Notas Classificacao_Notas  
0         0             0     0.218285                   F  
1         0             0     1.067764                   F  
2         0             0     3.018906                   B  
3         0             0     1.526868                   F  
4         0             0     1.058287                   F  
..      ...           ...          ...                 ...  
253       0             0     1.253452                   F  
254       0             0     0.534509                   F  
255       0             0     3.191046                   B  
256       1             0     2.815421                   C  
257       0             0     2.157122                   D  

[258 rows x 14 columns]

---

Quais as caracteristicas mais importantes para definir a Classificação das Notas?

---

In [70]:
# Características mais importantes para definir a Classificação das Notas

colunas = list(x_teste.columns)
importancia = pd.DataFrame(index=colunas, data=modelo_arvore_decisao.feature_importances_)
importancia = importancia * 100
print(importancia)

                                      0
Idade                          2.239083
Genero                         1.108586
Etnia                          1.949609
Escolaridade_Pais              2.229692
Horas_Estudo_Semanais          4.809006
Faltas                        23.406254
Aulas_Particulares             1.337815
Apoio_Pais                     2.784290
Atividades_Extracurriculares   1.119025
Esportes                       1.012912
Musica                         0.965540
Voluntariado                   0.768735
Media_Notas                   56.269454


---

As características mais importantes são as médias das notas com 56,2% e faltas com 23,4% e não menos importante as horas de estudos semanais com 4.8%.

---

**📈 Análise de Dados e Machine Learning**

**© 2025 - by Robson Silva - Programador Python e Analista de Dados.**

---